core/src/main/java/com/linecorp/armeria/common/logging/RequestContextExporter.java

```diff
@@ -235,15 +240,37 @@ public boolean containsBuiltIn(BuiltInProperty property) {
         requireNonNull(ctx, "ctx");
 
         final State state = state(ctx);
-        final RequestLog log = ctx.log().partial();
-        final int availabilities = log.availabilityStamp();
-        if (availabilities != state.availabilities) {
-            state.availabilities = availabilities;
-            export(state, ctx, log);
+        final RequestLogAccess log = ctx.log();
+
+        boolean needsUpdate;
+
+        // Needs to update if availabilityStamp has changed.
+        final int availabilityStamp = log.availabilityStamp();
+        if (state.availabilityStamp != availabilityStamp) {
+            state.availabilityStamp = availabilityStamp;
+            needsUpdate = true;
+        } else {
+            needsUpdate = false;
+        }
+
+        // Needs to update if any attributes have changed.
+        if (attrs != null) {
```

Comment:
`!needsUpdate && `

Reply:
Let me make it clear that this loop also updates `state.attrValues` while scanning.

Reply:
Ah :)